# Content 4.7 - Deriving New Variables

In [50]:
## Importing Data: 

import pandas as pd
import numpy as np
import os

path = r'C:\Users\cmfag\OneDrive\Desktop\Data Analyst Specialist\Achievement 4\4.3_orders_products\02 Data\Prepared Data'

In [51]:
#importing df_ords_prods_merge


df_ords_prods_merge = pd.read_pickle(os.path.join(path, 'ords_prods_merge.pkl'))

df_ords_prods_merge.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2539329        1    prior             1          2                  8   
2   2539329        1    prior             1          2                  8   
3   2539329        1    prior             1          2                  8   
4   2539329        1    prior             1          2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   

   Unnamed: 0                             product_name  aisle_id  \
0       195.0                                     Soda      77.0   
1     14084.0  Organic Unsweetened Vanilla Almond Milk      91.0   
2     12427.0                      Original Beef Jerky      23.0   
3     26089.0               Aged White Cheddar Popcorn      23.0   
4     26406.0         XL Pick-A-Size Paper Towel Rolls      54.0   

   department_id  prices merge_check  
0            7.0     9.0        both  
1           16.0    12.5        both  
2           19.0     4.4        both  
3           19.0     4.7        both  
4           17.0     1.0        both

## If-Statements

### If-Statements with User-Defined Functions

In [54]:
df = df_ords_prods_merge[:1000000]
#este subset incluye: everything from the beginning of the dataframe up to 1000000, es decir 1000000 rows.
df.shape

(1000000, 16)

In [55]:
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'


# esto quiere decir: 

#Rango de Bajo Costo (Low-range product):
   #Si el precio (prices) es menor o igual a 5, el producto se clasifica como "Low-range product".
   # Rango de Precio Medio (Mid-range product):
   #Si el precio está entre 5 y 15 (incluyendo 15), el producto se clasifica como "Mid-range product".
   # Rango de Alto Costo (High range):
   # Si el precio es mayor que 15, el producto se clasifica como "High range".
   # Datos Insuficientes (Not enough data):
   # Si el precio no cumple con ninguno de estos criterios, se etiqueta como "Not enough data" 
   #(esto es más un caso de cobertura para posibles datos faltantes o errores).

In [56]:
df['price_range'] = df.apply(price_label, axis=1)

C:\Users\cmfag\AppData\Local\Temp\ipykernel_32264\2848743372.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [57]:
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    672525
Low-range product    314109
High range            12413
Not enough data         953
Name: count, dtype: int64

In [58]:
df['prices'].max()



99999.0

### !!!!!  Preguntar al tutor - en el content, me sale que no hay ningun producto en la categoria high-range, pero a mi me salen 12413 precios. 

###  If-Statements with the loc() Function

#### - esta opción te permite hacer las misma etiqueta/segmentación en toda la base de dato, sin necesidad de generar un subset para evitar
####   problemas de memoria.

In [61]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

C:\Users\cmfag\AppData\Local\Temp\ipykernel_32264\1169838859.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [62]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [63]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [64]:
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    672525
Low-range product    314109
High range            12413
Not enough data         953
Name: count, dtype: int64

## If-Statements with For-Loops



In [84]:
# Check column names in df_ords_prods_merge
print(df_ords_prods_merge.columns)


Index(['order_id', 'user_id', 'eval_set', 'order_number', 'order_dow',
       'order_hour_of_day', 'days_since_prior_order', 'product_id',
       'add_to_cart_order', 'reordered', 'Unnamed: 0', 'product_name',
       'aisle_id', 'department_id', 'prices', 'merge_check'],
      dtype='object')


In [90]:
# Atención, por alguna razón mi df tiene order_dow y no orders_day_of_week


# Initialize an empty list to store the results
result = []

# Loop through each value in the correct column
for value in df_ords_prods_merge["order_dow"]:  
    if value == 0:
        result.append("Busiest day")
    elif value == 4:
        result.append("Least busy")
    else:
        result.append("Regularly busy")

# Assign result to a new column in the DataFrame
df_ords_prods_merge["day_label"] = result


In [88]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy

In [94]:
df_ords_prods_merge['busiest_day'] = result

In [98]:
df_ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22437988
Busiest day        6209808
Least busy         3787263
Name: count, dtype: int64

In [104]:
df_ords_prods_merge.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2539329        1    prior             1          2                  8   
2   2539329        1    prior             1          2                  8   
3   2539329        1    prior             1          2                  8   
4   2539329        1    prior             1          2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   

   Unnamed: 0                             product_name  aisle_id  \
0       195.0                                     Soda      77.0   
1     14084.0  Organic Unsweetened Vanilla Almond Milk      91.0   
2     12427.0                      Original Beef Jerky      23.0   
3     26089.0               Aged White Cheddar Popcorn      23.0   
4     26406.0         XL Pick-A-Size Paper Towel Rolls      54.0   

   department_id  prices merge_check       day_label     busiest_day  
0            7.0     9.0        both  Regularly busy  Regularly busy  
1           16.0    12.5        both  Regularly busy  Regularly busy  
2           19.0     4.4        both  Regularly busy  Regularly busy  
3           19.0     4.7        both  Regularly busy  Regularly busy  
4           17.0     1.0        both  Regularly busy  Regularly busy

In [108]:
df.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2539329        1    prior             1          2                  8   
2   2539329        1    prior             1          2                  8   
3   2539329        1    prior             1          2                  8   
4   2539329        1    prior             1          2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   

   Unnamed: 0                             product_name  aisle_id  \
0       195.0                                     Soda      77.0   
1     14084.0  Organic Unsweetened Vanilla Almond Milk      91.0   
2     12427.0                      Original Beef Jerky      23.0   
3     26089.0               Aged White Cheddar Popcorn      23.0   
4     26406.0         XL Pick-A-Size Paper Towel Rolls      54.0   

   department_id  prices merge_check        price_range    price_range_loc  
0            7.0     9.0        both  Mid-range product  Mid-range product  
1           16.0    12.5        both  Mid-range product  Mid-range product  
2           19.0     4.4        both  Low-range product  Low-range product  
3           19.0     4.7        both  Low-range product  Low-range product  
4           17.0     1.0        both  Low-range product  Low-range product